internal/expr/typeinfo.go

```diff
@@ -9,17 +9,31 @@ import (
 	"sync"
 )
 
+// typeMember should be implemented without pointer receivers as it is used as a
+// key in maps in some places.
 type typeMember interface {
 	outerType() reflect.Type
 	memberName() string
 }
 
+type primitiveType struct {
+	primitiveType reflect.Type
+}
+
+func (st primitiveType) outerType() reflect.Type {
+	return st.primitiveType
+}
+
+func (st primitiveType) memberName() string {
+	panic("internal error: memberName called on primitiveType")
```

Comment:
Here we are panicking and for other internal errors we are returning an error. The fact that this function cannot return an error/bool means that maybe it is the wrong abstraction for slices and primitive types.

Alternatively if we are changing the code to make better use of interfaces we may define a more specialized version of typeMember which satisfies typeMember and that has the `memberName` function. In the code we could check that the value satisfies this new interface.

Reply:
I agree, we need to replace this abstraction. Since we are refactoring the code anyway I am going to wait until the refactor is done to design the new interfaces. Hopefully then it will be clearer what is needed.